### PR TITLE
fix for cannot update x and x at the same time

### PR DESCRIPTION
**Pull request by marbemac**
If something is set then unset, or set to undefined, a modifier like this ends up being built:

```js
$set = { 'securitySchemes.oauth2.scopes': undefined }
$unset = { 'securitySchemes.oauth2.scopes': '' }
```

Which leads to this error:

```
"Cannot update 'securitySchemes.oauth2.scopes' and 'securitySchemes.oauth2.scopes' at the same time"
```

Wether or not this fix covers all possible edge cases I'm not sure? Unfortunately I'm not yet familiar enough with the Astronomy internals - just getting into it!

### CODE REVIEW

**Comment by jagi** [comment]
Could you show me exact example when it happens?

**Comment by marbemac** [comment]
Been playing around with this for about an hour. It is quite hard to reproduce, as it's buried in a massive import system we have. I believe the issue stems from this other issue, when nested properties are set as undefined.

In server only code:

```js
var model = Astro.Class.create({
  name: 'Test',
  collection: Tests,
  fields: {
    securitySchemes: {
      type: Object,
      default() {
        return {};
      },
    },
  },
});

var m = new model();
m.set('securitySchemes', { 
  oauth2: {
    name: 'oauth2',
    scopes: undefined, // NOTE THIS
  }
});
m.save(); // works ok for new model, fails for existing model

// incorrectly reports ["securitySchemes", "securitySchemes.oauth2", "securitySchemes.oauth2.scopes"]
m.getModified();

m.save() // produces error above
```

**Comment by marbemac** [comment]
In lib/modules/storage/utils/get_modified.js, if you recursively remove undefined properties from newDoc before comparing, it seems to fix the problem. ie something like this:

Note the "removeUndefined(newDoc)" line at the end of the file.

Whether or not this is the right place, or exactly right, I'm not 100% sure. Could probably make use of your raw_all / traverse stuff, which is already recursing through most of the props?

```js
import AstroClass from '../../../core/class.js';
import throwParseError from '../../../core/utils/throw_parse_error.js';
import rawAll from '../../fields/utils/raw_all.js';

function removeUndefined(obj) {
  return _.transform(obj, function (o, v, k) {
    if (_.isPlainObject(v)) {
      o[k] = removeUndefined(v);
    } else if (!_.isUndefined(v)) {
      o[k] = v;
    }
  });
};

function getModified(options = {}) {
  let {
    doc: newDoc,
    transient = false,
    immutable = false
  } = options;

  // Check if a document is an instance of the Astronomy class.
  if (!(newDoc instanceof AstroClass)) {
    throwParseError([{
        'module': 'storage'
      }, {
        'utility': 'getModified'
      }, {
        'argument': 'doc'
      },
      'Document has to be an instance of the Astronomy class'
    ]);
  }
  // Get class constructor from the document.
  let Class = newDoc.constructor;
  // Check if a document's class has associated collection.
  if (!Class.getCollection()) {
    throwParseError([{
        'module': 'storage'
      }, {
        'utility': 'getModified'
      }, {
        'argument': 'doc'
      },
      `Document's class does not have associated collection`
    ]);
  }
  // Get a document before modifications.
  let oldDoc = newDoc._isNew ? new Class() : Class.findOne(newDoc._id);
  // If there is no document before modifications that may mean that we are not
  // subscribed to the publication publishing given document or we modified the
  // _id of a document.
  if (!oldDoc) {
    throwParseError([{
        'module': 'storage'
      }, {
        'utility': 'getModified'
      },
      `Can not get a document before modifications. You are not subscribed ` +
      `to the publication publishing a "${Class.getName()}" document with ` +
      `the id "${newDoc._id}" or you have modified the "_id" field`
    ]);
  }

  let diff = function({
    newDoc,
    oldDoc,
    prefix = ''
  }) {
    let result = [];

    _.each(newDoc, function(newValue, fieldName) {
      let oldValue = oldDoc[fieldName];

      if (!EJSON.equals(oldValue, newValue)) {
        let nestedPrefix = (prefix && prefix + '.') + fieldName;
        result.push(nestedPrefix);

        // Compare two objects.
        if (_.isPlainObject(oldValue) && _.isPlainObject(newValue)) {

          result.push(...diff({
            oldDoc: oldValue,
            newDoc: newValue,
            prefix: nestedPrefix
          }));

        }
        // Compare two arrays.
        else if (_.isArray(oldValue) && _.isArray(newValue)) {

          let maxLength = Math.max(oldValue.length, newValue.length);
          _.each(_.range(maxLength), function(index) {
            let arrayPrefix = nestedPrefix + '.' + index;
            let oldElement = oldValue[index];
            let newElement = newValue[index];
            if (!EJSON.equals(oldElement, newElement)) {
              result.push(arrayPrefix);
              // If both array elements are object, then we perform diff between
              // them.
              if (_.isPlainObject(oldElement) && _.isPlainObject(
                  newElement)) {
                // Get a difference between elements.
                result.push(...diff({
                  oldDoc: oldElement,
                  newDoc: newElement,
                  prefix: arrayPrefix
                }));
              }
            }
          });

        }
      }
    });

    _.each(oldDoc, function(oldValue, fieldName) {
      let newValue = newDoc[fieldName];

      if (!EJSON.equals(oldValue, newValue)) {
        let nestedPrefix = (prefix && prefix + '.') + fieldName;
        if (newValue === undefined) {
          result.push(nestedPrefix);
        }
      }
    });

    return result;
  };

  return diff({
    // Get raw data from the docs.
    oldDoc: rawAll(oldDoc, {
      transient,
      immutable
    }),
    newDoc: rawAll(removeUndefined(newDoc), {
      transient,
      immutable
    })
  });
};

export default getModified;
```

**Comment by jagi** [comment]
Ok I think I know what can be a cause of the problem. You said that the problem occurs when values are set to undefined in a nested document. However, this PR will not solve the problem in the perfect way. I have to make some tests but in deed I may think of the situation where my current solution can cause some problems. I won't be able to make these tests and fix the problem in the next two weeks as I won't be available. But I will work on it when I'm back.

**Comment by marbemac** [comment]
After more digging, you're right and the above is not a solution, but I think I've finally figured it out. 

Basically, in raw_many.js you unravel astronomy classes, and if the resulting field is not another astronomy class, and not undefined, you include it. This means that fields defined as plain objects will not have their internal, undefined properties, removed. This is ok for updates, because you turn that into an $unset. However, for inserts, meteor/mongo turns those undefined props into "null". Thus, after insert, in getModified when you get the oldDoc, the value of the prop will be "null", while in the actual doc it is still undefined, causing getModified to not work.

Basically, I believe that on inserts you need to recursively remove properties with the value of undefined (even inside "blackbox" objects). Does this make sense?

**Comment by marbemac** [comment]
That fixes the error, but not the getModified issue (since the oldDoc will have no property, and the "newDoc" will have it set to undefined). To fix both, it might make sense to handle this at the set level, ie no setting of undefined properties if the document is new.

**Comment by jagi** [comment]
I think that the problem is in performing two passes for detecting $set and $unset modifiers. It would be best to perform it in one loop so that there would be no duplications of keys. I will investigate it. Thanks for noticing this problem.

**Comment by jagi** [comment]
UPDATE:
Ok I think that I've fixed a problem. Please, check out the newest version from GitHub and let me know if it fixes the error.

**Comment by marbemac** [comment]
Awesome! I was about to push some commits that fixed everything for me, but I'll give your solution a try first. Hold please :).

**Comment by marbemac** [comment]
Ok while your changes did fix the $unset bug, there are deeper, more subtle issues with how undefined works. I've fixed all the cases that I could find locally - hopefully it helps to identify the issue. I'm sure there are better ways to accomplish this with the existing Astronomy methods etc, but it's a reasonable start.

Here is a reproduction (on the server):

```js
'/astro/bug'() {
      var model = Astro.Class.create({
        name: 'Test',
        collection: Whatever,
        fields: {
          name: {
            type: String,
          },
          securitySchemes: {
            type: Object,
            default() {
              return {};
            },
          },
        },
      });

      var e = new model();
      e.name = Random.id();
      console.log('m0', e.getModified()); // correctly returns name
      e.save();

      e.set('securitySchemes', {
        oauth2: {
          name: 'oauth2',
          scopes: undefined,
          // scopes: '123',
        }
      });
      e.name = Random.id();
      console.log('m1', e.getModified()); // correctly returns securitySchemes and name

      e.save();
      console.log('m2', e.getModified()); // should return empty, but returns securitySchemes

      e.set('securitySchemes', {
        oauth2: {
          name: 'oauth2',
          scopes: undefined,
          // scopes: '123',
        }
      });
      e.save();

      console.log('m3', e.getModified()); // should return empty, but returns securitySchemes
    }
```

**Comment by jagi** [comment]
Are you sure that you're talking about the server code? Because it works in my tests. Could you create reproduction repository that I can clone?

**Comment by marbemac** [comment]
Sure, here you go: https://github.com/marbemac/astronomy-bug.

If you replace the astronomy package in there with my branch, getModified should work correctly.

**Comment by jagi** [comment]
Ok I've found the reason why it's happening. MongoDB does not accept setting fields` values to `undefined`. It converts such values to `null`. So you shouldn't use `undefined` because even in pure MongoDB it wouldn't work. Of course I could omit undefined values when saving and that way document on the server and document persisted in the MongoDB would be identical and `getModified()` would return empty array.

**Comment by marbemac** [comment]
Yup! That's exactly what I have been trying to explain in my posts above, and in the comments in my last commit in this PR :). This is coming from user uploaded structures, and while we could sprinkle logic everywhere to try and prevent this issue, ideally the "orm" can handle this. In fact, you do handle it, just not in blackbox objects, which causes all sorts of weird behavior (like described above).

It's more complicated than just a simple "omit undefined values when saving", because you are using undefined to allow users to $unset on updates.

Here are the three issues I identified in my last commit, with explanatory comments in the code. Without these changes, getModified will start return false positives in various scenarios, and cause future work with the model in question to start breaking (like the issue that started this entire thing).

1) 

For new models, undefined should not be allowed ever. I chose to do this in the set_one, but I'm not sure if set_one is called if the user directly sets a field ie myModel.name = "foo" instead of myModel.set('name', 'foo'). We always use .set, so this is ok for us for now.
https://github.com/stoplightio/meteor-astronomy/blob/v2/lib/modules/fields/utils/set_one.js#L19-L31

2)

After update, undefined values need to be removed from the document, since they have been $unset in the database.
https://github.com/stoplightio/meteor-astronomy/blob/v2/lib/modules/storage/utils/document_update.js#L65-L68

3) 

Never $set undefined, because mongo treats it differently.
https://github.com/stoplightio/meteor-astronomy/blob/v2/lib/modules/storage/utils/get_modifier.js#L136-L138

**Comment by jagi** [comment]
There is no problem with removing `undefined` values. It's just a matter when you do it. In the `getModifier()` method we need these values to generate `$unset` modifier - as you said. But after modifier is generated we can remove all `undefined` values from the `$set` modifier. I don't know how it will behave in the `$push` modifier so I have to check it out.

**Comment by jagi** [comment]
And yes ORM can handle that, so I will probably fix it. I'm adding it to my list of tasks. However I will probably take care of it in 2 weeks.

**Comment by marbemac** [comment]
I'm not sure I understand. Which parts of the 3 issues I described, and the pieces of my PR, do you think are not needed?

**Comment by jagi** [comment]
Changes to the `setOne` method are not necessary. The same with `documentUpdate`. The only change we need is `removeUndefined` in the `getModifier` method.

**Comment by jagi** [comment]
You're using the `set` method to set values but it's not obligatory, so removing `undefined` values is not required and even not recommended. When you do it, it would behave differently in these two situations:
```js
// 1
doc.set('nested', {a: 'a', b: undefined});
// 2
doc.nested = {a: 'a', b: undefined};
```
We can't do a lot if it goes about nested Objects without schema and `undefined` values in them. I know that you are probably not using schema for nested object, so that's why I'm writing it.

**Comment by marbemac** [comment]
Without all three changes, the test repo I made will still produce the issues. Perhaps I'm not explaining correctly:

If we don't remove undefined values after update, getModified will have issues. The model in memory has it set to undefined, and the model retrieved from the db in getModified (to compare for changes) will not have that field set. when getModified has issues, the rest of the system has issues, because it's how you build the modifiers and determine wether to save, etc.

This fix might be able to be consolidated into two pieces. 

1) The getModifier fix to never allow setting of undefined values. 
2) Clean all undefined values from the model after insert and update (instead of just update which is what I'm doing now).

Thoughts?

**Comment by marbemac** [comment]
BTW I wanted to add some tests, but couldn't figure out how to run the existing ones. How do you do that?

**Comment by jagi** [comment]
The `getModified` method is not used to generate modifiers, it's job of the `getModifiers` method. However, both methods are very similar. I may remove all undefined values but only from the fields of `Object` type because the problem occurs only in such fields.

If it goes about tests I'm using tinytest:
```sh
$ meteor test-packages ./ --release 1.3-rc.2 --port 4000
```

**Comment by marbemac** [comment]
I understand, the same principles apply to both though, and the way it works now returns incorrect values for both getModifiers and getModified.

Yup, this is only an issue for blackbox objects. When people use these objects, it will usually be because they're storing arbitrary data from some source (otherwise, if structure is known, they might as well define a schema). When storing arbitrary data, undefined values might be in there, and if they are, will cause subtle bugs like we have been running into.

Thanks on the tests.

**Comment by jagi** [comment]
Yes right. I will try to cover all cases. I will definitely use your `removeUndefined` method. The way how you've implemented `getModifier` is nice. The only problem I have is the `documentUpdate` method which just sets all fields in a document which is not perfect solution. It's better to get only `Object` fields and perform such operation. I'm not quite sure about `getModified` method, so I will have to make some tests.

If you want to work on it and prepare a new PR I will be glad. And you said that you would like to create tests, that would be awesome too.

**Comment by marbemac** [comment]
Ok cool, I'll clean it up and add tests in a separate PR :).

**Comment by jagi** [comment]
OK :)

**Comment by marbemac** [comment]
Hi @jagi, this bug is still not fixed in the latest v2, please see the reproduction below. The console.log statements describe the expected behavior, which you will see does not happen.

``` js
Meteor.methods({
    '/astro/bug'() {
      const model = Astro.Class.create({
        name: 'Test',
        collection: new Mongo.Collection('bugz'),
        fields: {
          name: {
            type: String,
          },
          securitySchemes: {
            type: Object,
            default() {
              return {};
            },
          },
        },
      });

      const e = new model();
      e.name = Random.id();
      console.log('name should be modified', e.getModified());
      e.save();

      e.set('securitySchemes', {
        oauth2: {
          name: 'oauth2',
          scopes: undefined,
          // scopes: '123',
        }
      });
      e.name = Random.id();
      console.log('name and securitySchemes should be modified', e.getModified());
      e.save();

      // Does not work as expected
      console.log('modified should be empty', e.getModified());
      e.set('securitySchemes', {
        oauth2: {
          name: 'oauth2',
          scopes: undefined,
          // scopes: '123',
        }
      });
      e.save();

      // Does not work as expected
      console.log('modified should be empty', e.getModified());
    }
});
```

**Comment by lukejagodzinski** [comment]
@marbemac yes in deed but it's related with a little bit different thing. However, I've managed to fix it. It's already on GitHub and will be available in 2.0. Thanks for noticing it!

**Comment by marbemac** [comment]
Hallelujah, fixed, thanks @jagi!
